import_obfuscation.go

```diff
@@ -0,0 +1,572 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/Binject/debug/goobj2"
+)
+
+// pkgInfo stores a parsed go archive/object file,
+// and the original path to which it was read from.
+type pkgInfo struct {
+	pkg  *goobj2.Package
+	path string
+}
+
+// dataType signifies whether the Data portion of a
+// goobj2.Sym is reflection data for an import path,
+// reflection data for a method of struct field, or
+// something else.
+type dataType uint8
+
+const (
+	other dataType = iota
+	importPath
+	namedata
+)
+
+// privateImports stores package paths and names that
+// match GOPRIVATE. privateNames are elements of the
+// paths in privatePaths, separated so that the shorter
+// names don't accidently match another import, such
+// as a stdlib package
+type privateImports struct {
+	privatePaths []string
+	privateNames []string
+}
+
+func obfuscateImports(objPath, importCfgPath string) (map[string]string, error) {
+	importCfg, err := goobj2.ParseImportCfg(importCfgPath)
+	if err != nil {
+		return nil, err
+	}
+	mainPkg, err := goobj2.Parse(objPath, "main", importCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing main objfile: %v", err)
+	}
+	privatePkgs := []pkgInfo{{mainPkg, objPath}}
+
+	// build list of imported packages that are private
+	for pkgPath, info := range importCfg {
+		if isPrivate(pkgPath) {
+			pkg, err := goobj2.Parse(info.Path, pkgPath, importCfg)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing objfile %s at %s: %v", pkgPath, info.Path, err)
+			}
+
+			privatePkgs = append(privatePkgs, pkgInfo{pkg, info.Path})
+		}
+	}
+
+	var sb strings.Builder
+	var buf bytes.Buffer
+	garbledImports := make(map[string]string)
+	for _, p := range privatePkgs {
+		// log.Printf("++ Obfuscating object file for %s ++", p.pkg.ImportPath)
+		for _, am := range p.pkg.ArchiveMembers {
+			// log.Printf("\t## Obfuscating archive member %s ##", am.ArchiveHeader.Name)
+
+			// skip objects that are not used by the linker, or that do not contain
+			// any Go symbol info
+			if am.IsCompilerObj() || am.IsDataObj() {
+				continue
+			}
+
+			// add all private import paths to a list to garble
+			var privImports privateImports
+			privImports.privatePaths, privImports.privateNames = explodeImportPath(p.pkg.ImportPath)
```

Comment:
Why does the `explodeImportPath` function not return a privateImports struct diretcly?

Reply:
Because the results are appended to `privImports`